tests/o2a_libs/test_functions.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for all functions module"""
+
+import unittest
+
+from parameterized import parameterized
+
+import o2a.o2a_libs.functions as functions
+from o2a.o2a_libs.el_wf_functions import _reverse_task_map
+
+
+class TestFunctions(unittest.TestCase):
+    @parameterized.expand(
+        [
+            ("foo", "bar", "foo"),
+            ("foo", "", "foo"),
+            ("foo", None, "foo"),
+            ("", "bar", "bar"),
+            (None, "bar", "bar"),
+            ("", "", ""),
+            (None, None, ""),
+        ]
+    )
+    def test_first_not_null(self, str_one, str_two, expected):
+        self.assertEqual(expected, functions.first_not_null(str_one, str_two))
+
+    @parameterized.expand(
+        [
+            ("foo", "bar", "foo ~ bar"),
+            ("foo", "", "foo"),
+            ("foo", None, "foo"),
+            ("", "bar", "bar"),
+            (None, "bar", "bar"),
+            (None, None, ""),
+        ]
+    )
+    def test_concat(self, str_one, str_two, expected):
+        self.assertEqual(expected, functions.concat(str_one, str_two))
+
+    @parameterized.expand(
+        [
+            ("foobar", "bar", "rab", "foorab"),
+            ("foobar", "", "rab", "foobar"),
+            ("foobar", None, "rab", "foobar"),
+            ("foobar", "bar", "", "foo"),
+            ("foobar", "bar", None, "foo"),
+            ("foobar", "[o]", "a", "faabar"),
+        ]
+    )
+    def test_replace_all(self, src_str, regex, replacement, expected):
+        self.assertEqual(expected, functions.replace_all(src_str, regex, replacement))
+
+    @parameterized.expand(
+        [
+            ("/a/b/,/c/b/,/c/d/", "ADD", ",", "/a/b/ADD,/c/b/ADD,/c/d/ADD"),
```

Comment:
Especially here I was scratching my head for a while trying to understand what's going on when I finally realised that expected values are on the right side.